File: Array/1642.Furthest_Building_You_Can_Reach/C++/1642.Furthest_Building_You_Can_Reach.cpp
class Solution {
public:
    int furthestBuilding(vector<int> & heights, int bricks, int ladders) 
    {
        //=======================================//
        // Input type:                           //
        //   - vector<int> ref. (heights vector) //
        //   - int (bricks quantity)             //
        //   - int (ladders quantity)            //
        // Return type:                          //
        //   - int (result maximun index)        //
        //=======================================//

        /*Initialize*/
        ///// Length of heights vector /////
        int len_heights = heights.size();

        ///// Record heap, priority-queue /////
        priority_queue<int, vector<int>, greater<int>> record_pq;


        /*Heap, Priority-queue based loop traversal*/
        for (int heights_idx = 1; (heights_idx < len_heights); (heights_idx++)) //Whole
        {
            ///// Check if the current heights difference matched conditions or not /////
            if (heights[heights_idx] <= heights[(heights_idx - 1)]) { continue; }

            record_pq.push((heights[heights_idx] - heights[(heights_idx - 1)])); //Keep updating, recording

            ///// Check if the current ladders enough or not /////
            if (record_pq.size() > ladders)
            {
                ///// Check if the current bricks enough or not /////
                if (record_pq.top() > bricks) { return (heights_idx - 1); }

                bricks -= record_pq.top(); //Keep updating, reducing

                record_pq.pop(); //Keep updating, popped
            }
            else { ; }

        } //Whole

        return (len_heights - 1);
    }
};

File: Array/1642.Furthest_Building_You_Can_Reach/Python/1642.Furthest_Building_You_Can_Reach.py
class Solution(object):
    def furthestBuilding(self, heights, bricks, ladders):
        """
        :type heights: List[int]
        :type bricks: int
        :type ladders: int
        :rtype: int
        """
        #============================================#
        # Heap/Priority-queue based traversal method #
        #============================================#

        ############
        #Initialize
        ##### Length of heights array #####
        len_heights = len(heights)

        ##### Record heap/priority-queue #####
        record_pq = []

        heapify(record_pq)


        ##########################################
        #Heap/Priority-queue based loop traversal
        for heights_idx in range(1, len_heights):

            ##### Check if the current heights difference matched conditions or not #####
            if (heights[heights_idx] <= heights[(heights_idx - 1)]):
                continue

            heappush(record_pq, (heights[heights_idx] - heights[(heights_idx - 1)])) #Keep updating/recording

            ##### Check if the current used ladders enough or not #####
            if (len(record_pq) > ladders):

                ##### Check if the current used bricks enough or not #####
                if (record_pq[0] > bricks):
                    return (heights_idx - 1)
                
                bricks -= record_pq[0] #Keep updating/reducing

                heappop(record_pq) #Keep updating/popped

            else:
                pass

        return (len_heights - 1)
